MastForest maximum node length invariant

sergerad opened the pull request:
Closes #1392

* Enforce invariant that MastForest node length is < 2^30
* Return errors from add_node() and enforce_node()
* Map errors to AssemblyError

sergerad commented:
@plafer LMK if this makes sense.

And are we happy panicking in `add_node()`? Or do we want to return error and refactor usage?

plafer commented:
@sergerad Thank you for taking this on, it is super appreciated!

The current implementation looks great; however, could you rebase it on the `plafer-mast-forest-serialization` branch? We can then wait for that PR to be merged before we merge this one. When I created #1392, I assumed the PR would be merged by the time anyone started working on it 😅.

Hence, the current implementation can stay the same, and the new `MastNodeId::from_u32_safe()` could be modified to use the new `TryFrom` that you implemented.

> And are we happy panicking in add_node()? Or do we want to return error and refactor usage?

If you are willing to do the work, returning an error would be much better as it's definitely the direction we want to go in. This would also require a similar change to `MastForestBuilder::ensure_node()`.

plafer commented:
Actually, thinking about it a bit more, I think it makes more sense for `MastForest` to know about the $2^{30}$ limit on number of nodes as opposed to the `MastNodeId` (a constraint that ultimately comes from the `MastForest` serialization format). So:

- `MastNodeId::from_u32_safe()` remains unchanged
- `MastForest::add_node()` contains the $2^{30}$ node limit check, and returns an error instead of panicking

sergerad commented:
Thanks very much for that guidance @plafer. I will make those changes including the removal of the panic.

sergerad commented:
@plafer I have added the node count invariant logic to MastForest and updated the relevant fns to return errors.

Most of the diffs are due to adding unwrap() in test files. All other files should be handling the new errors, not unwrapping.

I have done my best to handle those errors appropriately and map to AssemblyError. Let me know if and how to change the approach I have there with MastForestError.

bobbinth commented on core/src/mast/serialization/mod.rs:
```diff
@@ -120,7 +120,9 @@ impl Deserializable for MastForest {
-                mast_forest.add_node(node);
+                mast_forest
+                    .add_node(node)
+                    .map_err(|e| DeserializationError::UnknownError(e.to_string()))?;
```
The only reason this could fail is because we've serialized more than 2^30 nodes, right? If so, I'd probably use the `InvalidValue` error variant.

plafer replied:
Let's use `DeserializationError::InvalidValue` instead, and maybe with an error string that looks like `format!("failed to add node to MAST forest while deserializing: {}", e.to_string())`

bobbinth commented on assembly/src/assembler/mod.rs:
```diff
@@ -796,7 +796,7 @@ impl Assembler {
                             MastNode::new_split(then_blk, else_blk, mast_forest_builder.forest());
 
-                        mast_forest_builder.ensure_node(split_node)
+                        mast_forest_builder.ensure_node(split_node).map_err(AssemblyError::from)?
```
Is `.map_err()` here (and in other similar places) necessary? I would have thought that the compiler would be able to infer the conversion automatically and just `mast_forest_builder.ensure_node(split_node)?` would have been enough.

sergerad replied:
Updated. Just noting that the conversion required MastForestError to implement the miette::Diagnostic trait to for this. So the core crate now imports miette.

bobbinth replied:
Ah! In that case, I'd probably prefer to keep the extra `map_err()` here and avoid an additional import.

bobbinth commented on assembly/src/assembler/instruction/procedures.rs:
```diff
@@ -91,37 +91,43 @@ impl Assembler {
-                    mast_forest_builder.find_procedure_root(mast_root).unwrap_or_else(|| {
+                    if let Some(root) = mast_forest_builder.find_procedure_root(mast_root) {
+                        root
+                    } else {
                         // If the MAST root called isn't known to us, make it an external
                         // reference.
                         let external_node = MastNode::new_external(mast_root);
-                        mast_forest_builder.ensure_node(external_node)
-                    })
+                        mast_forest_builder.ensure_node(external_node)?
+                    }
```
I would probably write this as:
```rust
match mast_forest_builder.find_procedure_root(mast_root) {
    None => {
        // If the MAST root called isn't known to us, make it an external
        // reference.
        let external_node = MastNode::new_external(mast_root);
        mast_forest_builder.ensure_node(external_node)?
    }
    Some(root) => root
}
```
To keep it consistent with how we have things below.

bobbinth commented on assembly/src/assembler/basic_block_builder.rs:
```diff
@@ -135,7 +135,7 @@ impl BasicBlockBuilder {
 
             let basic_block_node = MastNode::new_basic_block_with_decorators(ops, decorators);
-            let basic_block_node_id = mast_forest_builder.ensure_node(basic_block_node);
+            let basic_block_node_id = mast_forest_builder.ensure_node(basic_block_node).unwrap();
```
Shouldn't this `unwrap()` also be an error? It does complicate the return type a bit - but I'm not sure if there is a good way to get around this.

sergerad replied:
Good catch, I missed this!

bobbinth reviewed (commented):
Looks good! Thank you for tackling this! I left a few small comments inline.

plafer reviewed (commented):
Looks good! Just left a few minor comments

sergerad commented on core/src/mast/mod.rs:
```diff
@@ -100,20 +110,21 @@ impl MastForest {
 
 /// Mutators
 impl MastForest {
+    /// The maximum number of nodes that can be stored in a single MAST forest.
+    const MAX_NODES: usize = 1 << 30;
```
Yes that's right - `1 << 30` is the non inclusive upper bound, which doesn't make sense as a value for `MAX_NODES`

sergerad commented on assembly/src/assembler/basic_block_builder.rs:
```diff
@@ -135,7 +135,7 @@ impl BasicBlockBuilder {
 
             let basic_block_node = MastNode::new_basic_block_with_decorators(ops, decorators);
-            let basic_block_node_id = mast_forest_builder.ensure_node(basic_block_node);
+            let basic_block_node_id = mast_forest_builder.ensure_node(basic_block_node).unwrap();
```
Yep, missed this one - treated it as a test file

sergerad commented:
Think I have covered all the current comments. Thanks!

plafer reviewed (approved):
LGTM! (let's fix clippy)

plafer commented:
Merged #1370, and changed the target branch to `next`

sergerad commented:
Fixed clippy and changelog 👍

plafer reviewed (approved):
🙏

bobbinth commented on core/Cargo.toml:
I think addressing https://github.com/0xPolygonMiden/miden-vm/pull/1394#discussion_r1683579436 should also revert changes to this file.

bobbinth reviewed (commented):
Looks good! Thank you! I left one small comment inline.

bobbinth reviewed (approved):
All looks good! Thank you!
